### PR TITLE
[Test Infra] Enable test infra to run on multiple tiles and random tensor input shape (First iteration)

### PR DESCRIPTION
**Pull request by ldjurovicTT**
### Ticket
#235 

### Problem description
Upgrade test infra to support multiple tiles

### What's changed
Addresses for L1 operands and result are generated in build.h ( still needs a bit of polishing ), some of tests are uplifted to work on custom tensor input shape ( for example [32,64] ). Based on that tile_cnt is generated and data is stored in L1 in contignuous manner aware of data format and it's length.

Golden generation for tilize ( untilize in progress ) is now treated as tilize/untilize_block which operate on multiple tiles of data.

All cpp tests changed to accept parameters from build.h

Fill_dest is treated as an edge case of multiple_tiles these ( where it gets filld with 16 (16 bit formats) or 8 (fp32 int32) tiles) so is removed as redundant.

Tests that are fully converted to new approach:
- eltwise unary datacopy
- eltwise unary sfpu
- Unpack tilize
- multiple tiles eltwise

### Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Checklist

- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/all-post-commit-workflows.yaml) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI passes (if applicable)

### CODE REVIEW

File: tests/python_tests/helpers/tilize_untilize.py [hunk 1]
```diff
@@ -7,6 +7,55 @@
 from .format_config import DataFormat
 
 
+def tilize_block(input_tensor, dimensions, stimuli_format=DataFormat.Float16_b):
```
Comment (fvranicTT) [inline tests/python_tests/helpers/tilize_untilize.py:10]: Wdyt about?
```python
def tilize_block(input_tensor, dimensions, stimuli_format=DataFormat.Float16_b):
    if input_tensor.numel() != dimensions[0] * dimensions[1]:
        raise ValueError(
            f"Cannot reshape tensor of size {input_tensor.numel()} to shape {dimensions}."
        )
    
    input_reshaped = input_tensor.view(dimensions[0], dimensions[1])
    if input_reshaped.ndim != 2:
        raise ValueError(
            f"Expected a 2D tensor for tilize_block, got shape {input_tensor.shape}"
        )
    
    rows, cols = input_reshaped.shape
    if rows % 32 != 0 or cols % 32 != 0:
        raise ValueError(
            f"Input tensor dimensions must be divisible by 32. Got shape: {input_tensor.shape}"
        )
    
    # Calculate number of blocks in each dimension
    row_blocks = rows // 32
    col_blocks = cols // 32
    total_blocks = row_blocks * col_blocks

    blocked_tensor = input_reshaped.reshape(row_blocks, 32, col_blocks, 32)
    
    # Permute to get blocks in the right order: (row_blocks, col_blocks, 32, 32)
    blocked_tensor = blocked_tensor.permute(0, 2, 1, 3)
    
    # Reshape to get all blocks as sequential entities: (total_blocks, 32, 32)
    all_blocks = blocked_tensor.reshape(total_blocks, 32, 32)

    flat_blocks = all_blocks.reshape(total_blocks, -1)

    tilized_blocks = torch.stack([
        tilize(block, stimuli_format=stimuli_format) 
        for block in flat_blocks
    ])
    
    # Reshape tilized blocks back to original dimensions
    tilized_output = tilized_blocks.flatten().reshape(rows, cols).to(format_dict[stimuli_format])
    
    return tilized_output
```

File: tests/python_tests/helpers/tilize_untilize.py [hunk 2]
```diff
@@ -39,3 +93,42 @@ def untilize(tilized_tensor, stimuli_format=DataFormat.Float16_b):
     original_tensor = torch.cat((top, bottom), dim=0).view(1024)
 
     return original_tensor.to(dtype=format_dict[stimuli_format])
+
+
+def untilize_block(
```
Comment (fvranicTT) [inline tests/python_tests/helpers/tilize_untilize.py:98]: Also:
```python
def untilize_block(
    input_tensor, stimuli_format=DataFormat.Float16_b, dimensions=[32, 32]
):
    """Optimized function to untilize blocks of data.
    
    Args:
        input_tensor: Input tensor to be untilized
        stimuli_format: Data format
        dimensions: Target dimensions for the output
    
    Returns:
        Untilized tensor with specified dimensions and data format
    """
    if input_tensor.numel() != dimensions[0] * dimensions[1]:
        raise ValueError(
            f"Cannot reshape tensor of size {input_tensor.numel()} to shape {dimensions}."
        )
    
    # Calculate number of 32x32 tiles
    rows, cols = dimensions
    row_blocks = rows // 32
    col_blocks = cols // 32
    total_blocks = row_blocks * col_blocks

    if rows % 32 != 0 or cols % 32 != 0:
        raise ValueError(
            f"Dimensions must be divisible by 32. Got dimensions: {dimensions}"
        )

    # Reshape input to have one block per 1024 elements
    input_reshaped = input_tensor.reshape(total_blocks, 1024)
    
    untilized_blocks = torch.stack([
        untilize(block) for block in input_reshaped
    ])

    output = untilized_blocks.reshape(row_blocks, col_blocks, 32, 32)

    # Then permute and reshape to get the final dimensions
    output = output.permute(0, 2, 1, 3).reshape(rows, cols)

    return output.to(dtype=format_dict[stimuli_format])
```

File: tests/python_tests/helpers/test_config.py [hunk 3]
```diff
@@ -127,28 +133,63 @@ def generate_build_header(
...
-    if test_config.get("testname") == "multiple_tiles_eltwise_test":
-        header_content.extend(
-            [
-                "",
-                "// Multiple tiles test configuration",
-                "#define MULTIPLE_OPS",
-                f"#define KERN_CNT {test_config.get('kern_cnt', 1)}",
-            ]
+    # Unpack an result buffer addresses arrrays generations
+    buffer_A_address = read_word_from_device("0,0", L1_buffer_locations.srcA.value)
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tests/python_tests/helpers/test_config.py:144]: This reads beginning address of buffer_A which is hardcoded and according to that computes where buffers should be written in L1

File: tests/sources/eltwise_unary_datacopy_test.cpp [hunk 4]
```diff
@@ -23,11 +23,15 @@ uint32_t math_sync_tile_dst_index = 0;
 
 void run_kernel()
 {
-    volatile uint32_t* const buffer_A = reinterpret_cast<volatile uint32_t*>(0x1a000);
+    // volatile uint32_t* const buffer_A = reinterpret_cast<volatile uint32_t*>(0x1a000);
```
Comment (fvranicTT) [inline tests/sources/eltwise_unary_datacopy_test.cpp:26]: to be removed

File: tests/sources/tilize_calculate_untilize_L1.cpp [hunk 5]
```diff
@@ -14,11 +14,10 @@ uint32_t unp_cfg_context          = 0;
 uint32_t pack_sync_tile_dst_ptr   = 0;
 uint32_t math_sync_tile_dst_index = 0;
 
-volatile uint32_t* const buffer_A = reinterpret_cast<volatile uint32_t*>(0x1a000);
-volatile uint32_t* const buffer_B = reinterpret_cast<volatile uint32_t*>(0x1b000);
+// TODO: CLEANUP
```
Comment (fvranicTT) [inline tests/sources/tilize_calculate_untilize_L1.cpp:17]: Can you link the issue number to this todo? We shouldn't have bare todos in our code, as they don't carry enough information what has to be done.

File: tests/python_tests/test_matmul.py [hunk 6]
```diff
@@ -71,24 +71,59 @@
...
+        formats.output_format,
+        math_fidelity,
+        input_dimensions=input_dimensions,
+    )
+    golden_tensor = tilize_block(
+        golden_tensor, dimensions=input_dimensions, stimuli_format=formats.output_format
+    ).to(torch_format)
+    golden_tensor = golden_tensor.flatten()
+
+    print(
```
Comment (fvranicTT) [inline tests/python_tests/test_matmul.py:98]: no prints please :)

File: tests/python_tests/helpers/format_arg_mapping.py [hunk 7]
```diff
@@ -87,3 +87,18 @@ class Mailbox(Enum):
...
     Math = 0x19FF8
     Packer = 0x19FF4
+
+
+format_tile_sizes = {
+    DataFormat.Bfp8_b: 1088,
+    DataFormat.Float16: 2048,
+    DataFormat.Float16_b: 2048,
+    DataFormat.Float32: 4096,
+    DataFormat.Int32: 4096,
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tests/python_tests/helpers/format_arg_mapping.py:97]: Let's leave this for another PR that will do cleanup of these kind of things. I will open an issue to track this. This PR is for pure functionality

[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tests/python_tests/helpers/format_arg_mapping.py:97]: #410

File: tests/python_tests/helpers/device.py [hunk 8]
```diff
@@ -34,26 +38,26 @@
...
 )
-from .utils import calculate_read_byte_count
 
 MAX_READ_BYTE_SIZE_16BIT = 2048
 
 
 def collect_results(
     formats: FormatConfig,
-    tensor_size: int,
+    tile_cnt: int,
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tests/python_tests/helpers/device.py:52]: Changed in all tests and python helper files. tile_cnt is now used only as temp variable in tests and is passed explicitly as tile_count=tile_cnt

File: tests/python_tests/helpers/test_config.py [hunk 9]
```diff
@@ -127,28 +133,63 @@ def generate_build_header(
...
+    buffer_B_address = read_word_from_device("0,0", L1_buffer_locations.srcB.value)
+    res_buffer_address = read_word_from_device("0,0", L1_buffer_locations.Res.value)
+
+    buffer_A_array = []
+    buffer_B_array = []
+    buffer_res_array = []
+
+    for i in range(tile_cnt):
```
[the remark this answers is not in the document]
Comment (ldjurovicTT) [reply tests/python_tests/helpers/test_config.py]: #410
